File: schema/examples_load.test.ts
import { describe, it, expect } from "vitest"
import { readFileSync, readdirSync, existsSync } from "fs"
import { join } from "path"
import type { DatasourceConfigSchema, ConfigField } from "./schema"
import { validateSchema } from "./validate"

/**
 * Cross-language contract test: loads example JSON files and validates
 * them with the TypeScript validateSchema(). The same files are validated
 * by Go tests in examples_load_test.go, ensuring both implementations
 * agree on what constitutes a valid schema.
 */

const examplesDir = join(__dirname, "examples")

function loadExample(filename: string): DatasourceConfigSchema {
    const data = readFileSync(join(examplesDir, filename), "utf-8")
    return JSON.parse(data) as DatasourceConfigSchema
}

function exampleFiles(): string[] {
    return readdirSync(examplesDir).filter((f) => f.endsWith(".schema.json"))
}

describe("example files — TS validation", () => {
    const cases: Array<{
        file: string
        description: string
        fieldCount: number
    }> = [
            {
                file: "simple-url.schema.json",
                description: "Minimal schema: single URL field with pattern validation",
                fieldCount: 1,
            },
            {
                file: "bearer-token.schema.json",
                description: "Auth schema: URL + auth method select + secure bearer token",
                fieldCount: 3,
            },
            {
                file: "indexed-headers.schema.json",
                description: "Storage mapping: array with indexedPair mapping",
                fieldCount: 4,
            },
            {
                file: "virtual-auth.schema.json",
                description: "Virtual fields: basic auth with computed field",
                fieldCount: 5,
            },
            {
                file: "array-of-objects.schema.json",
                description: "Array of objects: trace-to-metrics queries",
                fieldCount: 4,
            },
        ]

    for (const tc of cases) {
        it(`${tc.file}: ${tc.description}`, () => {
            const schema = loadExample(tc.file)

            // Validate the full schema
            const errors = validateSchema(schema)
            expect(errors).toHaveLength(0)

            // Verify root fields
            expect(schema.schemaVersion).toBeTruthy()
            expect(schema.pluginType).toBeTruthy()
            expect(schema.pluginName).toBeTruthy()
        })
    }
})

describe("example files — TS round-trip", () => {
    const files = exampleFiles()

    for (const file of files) {
        it(`${file}: survives JSON.stringify/parse and validates`, () => {
            const original = loadExample(file)
            const errors = validateSchema(original)
            expect(errors).toHaveLength(0)

            // Round-trip
            const json = JSON.stringify(original)
            const decoded: DatasourceConfigSchema = JSON.parse(json)
            expect(validateSchema(decoded)).toHaveLength(0)

            // Key properties survive
            expect(decoded.schemaVersion).toBe(original.schemaVersion)
            expect(decoded.pluginType).toBe(original.pluginType)
            expect(decoded.fields).toHaveLength(original.fields.length)
        })
    }
})

describe("example files — all discovered files validate", () => {
    it("every .schema.json file in examples/ passes TS validation", () => {
        const files = exampleFiles()
        expect(files.length).toBeGreaterThanOrEqual(5)

        for (const file of files) {
            const schema = loadExample(file)
            const errors = validateSchema(schema)
            expect(errors, `${file} should validate cleanly`).toHaveLength(0)
        }
    })
})

// ============================================================
// Datasource config.schema.json files (src/*/)
// ============================================================

const srcDir = join(__dirname, "..", "src")

function datasourceSchemaFiles(): string[] {
    if (!existsSync(srcDir)) {
        return []
    }
    return readdirSync(srcDir, { withFileTypes: true })
        .filter((d) => d.isDirectory() && existsSync(join(srcDir, d.name, "config.schema.json")))
        .map((d) => d.name)
}

function loadDatasourceSchema(dir: string): DatasourceConfigSchema {
    const data = readFileSync(join(srcDir, dir, "config.schema.json"), "utf-8")
    return JSON.parse(data) as DatasourceConfigSchema
}

/** Collect all fields including nested item fields */
function collectAllFields(fields: ConfigField[]): Array<{ path: string; field: ConfigField }> {
    const result: Array<{ path: string; field: ConfigField }> = []
    for (const f of fields) {
        result.push({ path: f.id, field: f })
        if (f.item?.fields) {
            for (const sub of f.item.fields) {
                result.push({ path: `${f.id}.item.${sub.id}`, field: sub })
            }
        }
    }
    return result
}

describe("datasource schemas — validation", () => {
    const dirs = datasourceSchemaFiles()

    it("should find at least 60 datasource schemas", () => {
        expect(dirs.length).toBeGreaterThanOrEqual(60)
    })

    for (const dir of dirs) {
        it(`${dir}: validates cleanly`, () => {
            const schema = loadDatasourceSchema(dir)
            const errors = validateSchema(schema)
            expect(errors, `${dir}/config.schema.json validation errors`).toHaveLength(0)
        })
    }
})

describe("datasource schemas — every field must have label and description", () => {
    const dirs = datasourceSchemaFiles()

    for (const dir of dirs) {
        it(`${dir}: all fields have label and description`, () => {
            const schema = loadDatasourceSchema(dir)
            const allFields = collectAllFields(schema.fields)
            const missing: string[] = []

            for (const { path, field } of allFields) {
                if (!field.label) {
                    missing.push(`${path}: missing label`)
                }
                if (!field.description) {
                    missing.push(`${path}: missing description`)
                }
            }

            expect(missing, `${dir}/config.schema.json fields missing label/description`).toHaveLength(0)
        })
    }
})

describe("datasource schemas — pluginType matches directory name", () => {
    const dirs = datasourceSchemaFiles()

    for (const dir of dirs) {
        it(`${dir}: pluginType matches directory`, () => {
            const schema = loadDatasourceSchema(dir)
            expect(schema.pluginType).toBe(dir)
        })
    }
})

File: schema/examples_load_test.go
package schema_test

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"github.com/grafana/dsconfig/schema"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// examplesDir returns the path to the examples directory relative to the
// test file. This allows `go test` to find examples from the schema/ dir.
func examplesDir() string {
	return filepath.Join("examples")
}

// loadExample reads a JSON example file, unmarshals it into a
// DatasourceConfigSchema, and returns it. Fails the test on any error.
func loadExample(t *testing.T, filename string) *schema.DatasourceConfigSchema {
	t.Helper()
	path := filepath.Join(examplesDir(), filename)
	data, err := os.ReadFile(path)
	require.NoError(t, err, "failed to read example file: %s", filename)

	var s schema.DatasourceConfigSchema
	require.NoError(t, json.Unmarshal(data, &s), "failed to unmarshal: %s", filename)
	return &s
}

// TestExampleFiles_GoValidation loads each example JSON file, unmarshals
// it into the Go schema struct, and runs Validate(). This ensures the
// examples are valid according to the Go contract and that Go's JSON
// tags correctly map to the schema field names.
func TestExampleFiles_GoValidation(t *testing.T) {
	examples := []struct {
		file        string
		description string
		fieldCount  int // expected number of field IDs (including item fields)
	}{
		{
			file:        "simple-url.schema.json",
			description: "Minimal schema: single URL field with pattern validation",
			fieldCount:  1,
		},
		{
			file:        "bearer-token.schema.json",
			description: "Auth schema: URL + auth method select + secure bearer token with conditional visibility",
			fieldCount:  3,
		},
		{
			file:        "indexed-headers.schema.json",
			description: "Storage mapping: array of header objects with indexedPair mapping to legacy jsonData/secureJsonData keys",
			fieldCount:  4, // url + httpHeaders + 2 item fields
		},
		{
			file:        "virtual-auth.schema.json",
			description: "Virtual fields: basic auth with computed virtual field and pair relationship",
			fieldCount:  5,
		},
		{
			file:        "array-of-objects.schema.json",
			description: "Array of objects: trace-to-metrics queries with item fields and textarea UI",
			fieldCount:  4, // url + tracesToMetrics + 2 item fields
		},
	}

	for _, tc := range examples {
		t.Run(tc.file, func(t *testing.T) {
			s := loadExample(t, tc.file)

			// Validate the full schema
			require.NoError(t, s.Validate(), "%s: %s", tc.file, tc.description)

			// Verify field ID count matches expectation
			ids, err := s.FieldIDs()
			require.NoError(t, err)
			assert.Len(t, ids, tc.fieldCount, "%s: expected %d field IDs", tc.file, tc.fieldCount)

			// Verify root fields were populated
			assert.NotEmpty(t, s.SchemaVersion)
			assert.NotEmpty(t, s.PluginType)
			assert.NotEmpty(t, s.PluginName)
		})
	}
}

// TestExampleFiles_GoRoundTrip verifies that each example survives
// Go unmarshal -> marshal -> unmarshal and still validates. This catches
// serialization issues where Go omitempty or field naming mismatches
// could silently drop data.
func TestExampleFiles_GoRoundTrip(t *testing.T) {
	files, err := filepath.Glob(filepath.Join(examplesDir(), "*.schema.json"))
	require.NoError(t, err)
	require.NotEmpty(t, files, "no example files found")

	for _, path := range files {
		name := filepath.Base(path)
		t.Run(name, func(t *testing.T) {
			data, err := os.ReadFile(path)
			require.NoError(t, err)

			// First pass: unmarshal + validate
			var original schema.DatasourceConfigSchema
			require.NoError(t, json.Unmarshal(data, &original))
			require.NoError(t, original.Validate())

			// Round-trip: marshal + unmarshal
			roundTripped, err := json.Marshal(&original)
			require.NoError(t, err)

			var decoded schema.DatasourceConfigSchema
			require.NoError(t, json.Unmarshal(roundTripped, &decoded))
			require.NoError(t, decoded.Validate())

			// Verify key properties survived
			assert.Equal(t, original.SchemaVersion, decoded.SchemaVersion)
			assert.Equal(t, original.PluginType, decoded.PluginType)
			assert.Len(t, decoded.Fields, len(original.Fields))
		})
	}
}
